src/scripts/GameConstants.ts

```diff
@@ -296,10 +296,22 @@ namespace GameConstants {
         return Math.floor(Math.random() * (max - min + 1)) + min;
     }
 
-    export function randomElement(array: any[]): any {
+    export function randomElement<T>(array: T[]): T {
         return array[GameConstants.randomIntBetween(0, array.length - 1)];
     }
 
+    // Return a random element from the array, with an exponential distribution
+    // The last element has a 1/ratio chance of being chosen, one before last is 1/(ratio^2), etc
+    // The logarithm is clipped up to 0, so the first two elements will have equal chance
+    export function expRandomElement<T>(array: T[], ratio: number): T {
+        const r = Math.random();
+        // Math.random() can be 0, and log(0) is NaN, so we turn it back to 0 if this happens.
+        const logr = (Math.log(r) || 0) / Math.log(ratio);
```

Comment:
`Math.log(0)` seems to return `-Infinity` so I don't think this will work as expected?
```suggestion
        // Math.random() can be 0, and log(0) is -Infinity, so we turn it back to 0 if this happens.
        const logr = (r ? Math.log(r) : 0) / Math.log(ratio);
```

Reply:
Huh, so it is...
In that case, we can just rely on the clipNumber taking care of it for us?

Reply:
Yeah, just becomes the last element of the array for `-Infinity`